File: Feature-Extraction-Dimensionality-Reduction/LDA.R
# Linear Discriminant Analysis - Feature Extraction
# Dimensionality Reduction

# importing dataset
dataset = read.csv('Wine.csv')

# splitting the dataset into training and test sets
library(caTools)
set.seed(1234)
split = sample.split(dataset$Customer_Segment, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# feature scaling
training_set[,1:13] = scale(training_set[,1:13])
test_set[,1:13] = scale(test_set[,1:13])

# applying lda
library(MASS)
lda_model = lda(formula = Customer_Segment ~., data = training_set)
training_set = as.data.frame(predict(lda_model, newdata = training_set))
test_set = as.data.frame(predict(lda_model, newdata = test_set))
training_set = training_set[c(5,6,1)]
test_set = test_set[c(5,6,1)]

# fitting svm to the training set
library(e1071)
classifier = svm(formula = class ~., 
                 data = training_set,
                 type = 'C-classification',
                 kernel = 'linear')

# predicitng test set results
y_pred = predict(classifier, newdata = test_set[,1:2])

# making confusion matrix
cm = table(test_set[,3], y_pred)

# visualising training set results
library(ElemStatLearn)
X1 = seq(min(training_set[,1])-1, max(training_set[,1])+1, by =0.01)
X2 = seq(min(training_set[,2]) -1, max(training_set[,2]) +1, by = 0.01)
grid_set = expand.grid(X1,X2)
colnames(grid_set) = c('x.LD1', 'x.LD2')
y_grid = predict(classifier, newdata = grid_set)
plot(training_set[,1:2], main = 'LDA',
     xlim = range(X1), ylim = range(X2), xlab = 'LD1', ylab = 'LD2')
contour(X1,X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse( y_grid ==2, 'white', ifelse(y_grid == 1, 'black', 'grey')))
points(training_set[,1:2], pch = 21, bg = ifelse(training_set[,3] == 2, 'red', 
                                                 ifelse(training_set[,3] == 1, 'yellow', 'blue')))


# visualising test set results
library(ElemStatLearn)
X1 = seq(min(test_set[,1]) -1, max(test_set[,1]) +1, by =0.01)
X2 = seq(min(test_set[,2]) -1, max(test_set[,2]) +1, by =0.01)
grid_set = expand.grid(X1,X2)
colnames(grid_set) = c('x.LD1','x.LD2')
y_grid = predict(classifier, newdata = grid_set)
plot(test_set[,1:2],main = 'LDA', xlab = 'LD1', ylab = 'LD2',
     xlim = range(X1), ylim = range(X2))
contour(X1,X2,matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse( y_grid ==2, 'orange', ifelse(y_grid == 1, 'green', 'pink')))
points(test_set[,1:2], pch = 21, bg = ifelse(test_set[,3] == 2, 'black', 
                                             ifelse(test_set[,3] == 1, 'yellow', 'blue')))

File: Feature-Extraction-Dimensionality-Reduction/LDA.py
# -*- coding: utf-8 -*-
"""
Created on Fri Dec  1 12:35:50 2017

@author: adhingra
"""
# Linear Discriminant Analysis - Feature Extraction
# Dimensionality Reduction

# importing libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# importing dataset
dataset = pd.read_csv('Wine.csv')
X = dataset.iloc[:,0:13].values
y = dataset.iloc[:,13].values

# splitting the dataset into training and test sets
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=0)

# feature scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

# analysing lda 
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
lda = LDA(n_components = 2)
X_train = lda.fit_transform(X_train,y_train)
X_test = lda.transform(X_test)

# fitting logistic regression model to the dataset
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train,y_train)

# predicting test set results
y_pred = classifier.predict(X_test)

# making confusion matrix, accuracy
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test,y_pred)
accuracy = accuracy_score(y_test,y_pred)

# visualing training set results
from matplotlib.colors import ListedColormap
X1,X2 = np.meshgrid(np.arange(start = X_train[:,0].min()-1, stop = X_train[:,0].max() +1, step = 0.01),
                    np.arange(start = X_train[:,1].min() -1, stop = X_train[:,1].max() +1, step = 0.01))
plt.contourf(X1,X2,classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.6, cmap = ListedColormap(('red', 'blue', 'green')))
plt.xlim(X1.min(),X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_train)):
    plt.scatter(X_train[y_train == j, 0], X_train[y_train == j, 1],
                c = ListedColormap(('white', 'yellow', 'brown'))(i), label =j)
plt.legend()
plt.title('PCA')
plt.xlabel('PCA1')
plt.ylabel('PCA2')
plt.show()
    
# visualising test set results
from matplotlib.colors import ListedColormap
X1,X2 = np.meshgrid(np.arange(start = X_test[:,0].min()-1, stop = X_test[:,0].max()+1, step =0.01),
                    np.arange(start = X_test[:,1].min()-1,stop = X_test[:,1].max()+1, step =0.01))
plt.contourf(X1,X2,classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             aplha = 0.6, cmap = ListedColormap(('brown', 'violet', 'black')))
plt.xlim(X1.min(), X2.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_test)):
    plt.scatter(X_test[y_test == j, 0], X_test[y_test == j, 1],
                c= ListedColormap(('white','red', 'yellow'))(i), label = j)
plt.legend()
plt.title('PCA')
plt.xlabel('PCA1')
plt.ylabel('PCA2')
plt.show()
